infra/vm_registry.go

```diff
@@ -0,0 +1,126 @@
+package infra
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+
+	"github.com/gravitational/trace"
+
+	"cloud.google.com/go/datastore"
+	log "github.com/sirupsen/logrus"
+)
+
+//gcsDatastoreVmRegistry stores VM registry records inside google datastore
+type dsVmRegistry struct {
+	client *datastore.Client
+	logger log.FieldLogger
+}
+
+type dsVmEntry struct {
+	// Enabled is whether this snapshot could be used for restore
+	Enabled bool `datastore:"enabled"`
+	// Cloud is which cloud provider the snapshot belongs to
+	Cloud string `datastore:"cloud"`
+	// Checkpoint is milestone in the cluster lifecycle when VM snapshot has been taken
+	Checkpoint string `datastore:"checkpoint"`
+	// Region is cloud region
+	Region string `datastore:"region"`
+	// ResourceGroup is resource group the VM snapshot belongs to
+	ResourceGroup string `datastore:"resource_group"`
+	// Param is JSON serialized list of parameters related to Checkpoint
+	Param string `datastore:"param,noindex"`
+	// Dir is directory where installer is located on the machine
+	InstallDir string `datastore:"dir"`
+	// K is entity key
+	K *datastore.Key `datastore:"__key__"`
+}
+
+const (
+	vmEntryKind = "VmImage"
+)
+
+func GCSDatastoreVmRegistry(ctx context.Context, projectID string, logger log.FieldLogger) (VmRegistry, error) {
+	client, err := datastore.NewClient(ctx, projectID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &dsVmRegistry{client, logger}, nil
+}
+
+// Locate returns VM image or trace.NotFound when none was found
+func (r *dsVmRegistry) Locate(ctx context.Context, cloud, checkpoint string, param interface{}) (*VmImage, error) {
+	var entries []dsVmEntry
+	q := datastore.NewQuery(vmEntryKind)
+	_, err := r.client.GetAll(ctx, q, &entries)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// serialize and deserialize back params for comparison
```

Comment:
Not sure about these back and forth, but `paramGeneralized` could be hidden inside of `compareJSON`.
`compareJSON` is not too descriptive - `compareParams` should tell more about the intent of the comparison (even though it is not satisfactory as well)